### PR TITLE
Lees-Edwards / lattice Boltzmann

### PR DESCRIPTION
**Pull request by bindgens1**
Fixes #

Description of changes:
We added the Lees-Edwards capability to the lattice Boltzmann GPU code (currently: two point coupling). Test cases are provided for  (i) transient velocity profile, (ii) boundary interpolation scheme for several cases and (iii) particle fluid coupling at the boundary. Documentation is already updated. 

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

**Comment by KaiSzuttor** [comment]
please check that it at least builds

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please add a description

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please add a description

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: if possible dont use ouput variables

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: this function does a lot at once, maybe split it into subfunctions

File: testsuite/lb_gpu_leesedwards_velocity_profile.py
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: superfluous comment

File: testsuite/lb_gpu_leesedwards_velocity_profile.py
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: typo

**Comment by bindgens1** [comment]
Will improve this and check with the other myconfigs. Lees-Edwards doesn't play nice with other features.

File: testsuite/lb_gpu_leesedwards_velocity_profile.py
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_velocity_profile.py]: remove newlines

File: testsuite/lb_gpu_leesedwards_particlecoupling.py
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_particlecoupling.py]: if you call it tolerance or similar youndont need the comment

File: testsuite/lb_gpu_leesedwards_gridoffset.py
Comment (KaiSzuttor) [inline testsuite/lb_gpu_leesedwards_gridoffset.py]: please combine all the tests into one testclass

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: testsuite/lb_gpu_leesedwards_gridoffset.py
Comment (bindgens1) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: How can I reset the LB to initial state?

Comment (fweik) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: You can't, as LB can not be disabled.

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: the interpolation should really not depend on any LE variable

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: same as above, this function should not depend on LE variables

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: testsuite/lb_gpu_leesedwards_gridoffset.py
Comment (KaiSzuttor) [reply testsuite/lb_gpu_leesedwards_gridoffset.py]: There is an open PR #1574 which should be merged and then used @fweik

File: src/core/lbgpu_cuda.cu
Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I could provide two functions (the old and the new one for LE) and distinguish their call with an #ifdef / #ifndef. But that would nearly duplicate the functions.

Comment (KaiSzuttor) [reply src/core/lbgpu_cuda.cu]: No, just split functionality. Create a new function for LE and if it shares a lot of the two_point_coupling function, split that part off into a new function and call that new function from the two_point coupling and your function.

**Comment by bindgens1** [comment]
Clang on gitlab doesn't configure: 

`-- Configuring incomplete, errors occurred!
See also "/builds/espressomd/espresso/build/CMakeFiles/CMakeOutput.log".
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
CUDART_LIBRARY`

Will merge the test cases next

**Comment by mkuron** [comment]
Just merge the current python branch to resolve the CUDART issue:
```
git pull upstream python
```

**Comment by bindgens1** [comment]
Thanks!

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: output parameters...

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: typo density

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: remove that line

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: typo lower, please describe each parameter

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: what happens here?

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: please name your functions according to what they do, i.e. calculate.... Please add doxygen documentation

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: Same as above, doxygen documentation and name of function should be verb.

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: remove empty line

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: int x, y, z;

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: Please don't use output variables

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu:2362]: doxygen documentation missing

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: cant you call the specifics to your algorithm from within the interpolation function and decide if it's called there?

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: doxygen documentation missing

Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: don't use output variables

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: doxygen documentation missing

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: clang-format -style=file -i

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: nullptr was correct

**Review by KaiSzuttor — CHANGES_REQUESTED** [review]

File: src/core/lbgpu_cuda.cu
Comment (uschille) [inline src/core/lbgpu_cuda.cu]: I suspect the `sqrt`s are wrong. The back transformation - see Eq. (2.66) in http://pubman.mpdl.mpg.de/pubman/item/escidoc:1423983/component/escidoc:1423982/Dissertation-Schiller.pdf needs to include the norm of the basis vectors which I do not see here. Cf. the statement `m[i] = (1./e[19][i])*mode[i];` in `lb_calc_n_from_modes(...)`. The quadrature weights `w[i]` should then enter without the `sqrt`.

Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I tried and it works much better. Will include that in the next commit! Thanks!

**Comment by bindgens1** [comment]
I will remove the prints and put all test cases in one file (with the box_l provided a the right spot) during one of the next commits.

File: build-lble/testsuite/lb_leesedwards.py
Comment (KaiSzuttor) [inline build-lble/testsuite/lb_leesedwards.py]: the gpu version is a feature that may not be compiled in, so you have to guard those statements that refer to the LBFluidGPU instance. Furthermore it's easier to set the system up in the class and only use a single testclass.

File: src/core/lbgpu_cuda.cu
Comment (bindgens1) [reply src/core/lbgpu_cuda.cu]: I don't really get what you mean. Are you referring to the xyz?

Comment (KaiSzuttor) [reply src/core/lbgpu_cuda.cu]: yes

File: src/core/lbgpu_cuda.cu
Comment (KaiSzuttor) [inline src/core/lbgpu_cuda.cu]: i meant to not use output variable at all and instead returning the value

**Comment by KaiSzuttor** [comment]
any news here?

**Comment by bindgens1** [comment]
As soon as the regular LE is merged I'll update this branch.

**Comment by fweik** [comment]
@bindgens1 my understanding is that we do not go forward with this, and instead directly into walberla, so this can be closed?

**Comment by bindgens1** [comment]
Sad but true :-/
If you want to remove the LB / LBGPU soon, it's not worth to put the effort in this project.

**Comment by fweik** [comment]
Sorry about that, so sad. As you might have noticed I'm also not happy with how this project progresses, let's just hope that we'll succeed eventually.

**Pull request closed without merging**
